tests/federation_acl_test.go

```diff
@@ -0,0 +1,142 @@
+package tests
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/match"
+	"github.com/matrix-org/complement/internal/must"
+)
+
+// Test for https://github.com/matrix-org/dendrite/issues/3004
+func TestACLs(t *testing.T) {
+	// 1. Prepare 3 or more servers. 1st will be room host, 2nd will be blocked with m.room.server_acl and 3rd server will be affected by this issue. 1st and 2nd servers don't have to be powered by dendrite.
+	deployment := Deploy(t, b.Blueprint{
+		Name: "federation_three_homeservers",
+		Homeservers: []b.Homeserver{
+			{
+				Name: "hs1",
+				Users: []b.User{
+					{
+						Localpart:   "alice",
+						DisplayName: "Alice",
+					},
+				},
+			},
+			{
+				Name: "hs2",
+				Users: []b.User{
+					{
+						Localpart:   "bob",
+						DisplayName: "Bob",
+					},
+				},
+			},
+			{
+				Name: "hs3",
+				Users: []b.User{
+					{
+						Localpart:   "charlie",
+						DisplayName: "Charlie",
+					},
+				},
+			},
+		},
+	})
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs2", "@bob:hs2")
+	charlie := deployment.Client(t, "hs3", "@charlie:hs3")
+
+	/*
+	   You can set m.room.server_acl again, to "fix" this.
+	*/
+
+	// 2. Create room on 1st server
+	roomID := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+	aliceSince := alice.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(alice.UserID, roomID))
+
+	// 3. Join this room from 2nd server
+	bob.JoinRoom(t, roomID, []string{"hs1"})
+	aliceSince = alice.MustSyncUntil(t, client.SyncReq{Since: aliceSince}, client.SyncJoinedTo(bob.UserID, roomID))
+	bobSince := bob.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(bob.UserID, roomID))
+
+	// create a different room used for a sentinel event
+	sentinelRoom := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+	aliceSince = alice.MustSyncUntil(t, client.SyncReq{Since: aliceSince}, client.SyncJoinedTo(alice.UserID, sentinelRoom))
+	bob.JoinRoom(t, sentinelRoom, []string{"hs1"})
+	charlie.JoinRoom(t, sentinelRoom, []string{"hs1"})
+	aliceSince = alice.MustSyncUntil(t, client.SyncReq{Since: aliceSince},
+		client.SyncJoinedTo(bob.UserID, sentinelRoom),
+		client.SyncJoinedTo(charlie.UserID, sentinelRoom),
+	)
+
+	// 4. Add deny rule, to block 2nd server from participating
+	stateKey := ""
+	eventID := alice.SendEventSynced(t, roomID, b.Event{
+		Type:     "m.room.server_acl",
+		Sender:   alice.UserID,
+		StateKey: &stateKey,
+		Content: map[string]interface{}{
+			"allow":             []string{"*"},
+			"allow_ip_literals": true,
+			"deny":              []string{"hs2"},
+		},
+	})
+	// wait for the ACL to show up on hs2
+	bob.MustSyncUntil(t, client.SyncReq{Since: bobSince}, client.SyncTimelineHasEventID(roomID, eventID))
+
+	// 5. Join from 3rd (dendrite) server.
```

Comment:
I'm guessing these comments are cribbed from the issue report? It'd be nice to keep Dendrite homeserver-agnostic where possible.

Reply:
```suggestion
	// 5. Join from 3rd server.
```